### DISCUSSION TITLE
有关 DTK 建议添加全局异常捕获和单例模式传参功能

### DISCUSSION THREAD

**Discussion by Wing-summer** [linuxdeepin/developer-center#3412] [Developer Discussion | 开发者讨论]
当我们开发程序的时候，我们并不能保证写的代码一定不会出错，但开发应用频繁的 try catch 太难维护，希望添加一个全局异常捕获接口，具体可以参考`qBreakpad`。这个对于稍大的程序一定是需要的，加上更可以方便帮助开发者更加精准的定位错误来源以及触发函数，而不必必须自己实现或者再引用第三方库。

DTK 提供了单例模式的实现，但是吧，它不能传参数，想要传参数还要自行实现，这个应该说对于稍大的应用程序也是个刚需，深度编辑器就自己实现的功能。

这个仅仅是个建议，不过我还是比较希望采纳的。
:smile:

**Comment by BLumia** [comment]
DTK 本身和 Qt 一样，是 exception-free 的，只使用 DTK 本身并不会造成写 `try catch` 的需求才对。至于提到的 `qBreakpad` 应该也并非规避写 `try catch` 而存在的，感觉更像是类似 DrKonqi 所存在的目的。我不清楚我的理解有没有误。

cc @zccrs

**Reply by zccrs** [reply]
对于异常的处理的确是这样的，DTK跟Qt一样不使用c++的异常系统。qBreakpad 我倒没看过，这是Qt的东西吗？

**Reply by Wing-summer** [reply]
既然是 exception-free 的，那就算了吧，设计理念在这就不强求了。qBreakpad 本质不是规避写 try catch 的，我只是想捕获我因考虑不周出的错误，你的理解是差不多的。

**Reply by BLumia** [reply]
补充：如果你的目的是方便本地调试而不是异常信息上报反馈的话，本地配合 systemd-coredump 的 `coredumpctl` 使用其实还是可以满足相应需求的 :)

**Reply by Wing-summer** [reply]
是异常的上报，就是别人运行我设计的软件出现异常，往往有时单纯的描述无法精确定位问题，配合出错的 dmp 会增加解决的可能性，感谢建议！

**Comment by zccrs** [comment]
单例模式的传参我记得是支持的

**Reply by zccrs** [reply]
`DGuiApplicationHelper::newProcessInstance(qint64 pid, const QStringList &arguments)` 会把新的进程实例的参数传递过来，可以试一下这个有没有用。不过不行可以提个 issue.

**Reply by Wing-summer** [reply]
我测试成功了，不过官方的文档和实例还是得完善。不知道是不是我没有找到例子和文档比较全的网站，如果有点话能不能麻烦给一下。:smile:

**Reply by BLumia** [reply]
这里有一份根据目前代码生成的文档 https://linuxdeepin.codeberg.page/dtkgui/@docs~master/classDtk_1_1Gui_1_1DGuiApplicationHelper.html （总的在[这里](https://linuxdeepin.codeberg.page/)，注：非官方维护的页面，官方维护的文档在[这里](https://docs.deepin.org/)，但官网这个页面的文档相对旧一些）

目前 DTK 文档确实不太全，这个文档问题我记录下，如果你有别的文档问题也可以另开个 discussion 或者 issue 反馈哈。

**Reply by BLumia** [reply]
相关文档的更新：

- https://github.com/linuxdeepin/dtkgui/pull/60

（ps. 有兴趣的话可以一起来完善）

**Comment by zccrs** [comment]
看了下你的项目，能跟一位论坛用户对上号了🐶。如果有兴趣的话可以试试 DTK 的 qml 版本： https://github.com/linuxdeepin/dtkdeclarative ，它的发展前景更好（如果有动画和一些特效的需求的话它更合适），不过 dtkwidget 比它更成熟些是自然的。

**Reply by zccrs** [reply]
有 https://github.com/linuxdeepin/deepin-tweak/pull/1 这个项目正在趟路。

**Reply by Wing-summer** [reply]
哈哈，那个论坛用户应该就是我没错了。看到头像一样，昵称就差个大小写和多个-就是我了。
